projects/RabbitMQ.Client/client/impl/SessionBase.cs

```diff
@@ -159,31 +159,48 @@ public void Notify()
             OnSessionShutdown(CloseReason);
         }
 
-        public virtual void Transmit(in OutgoingCommand cmd)
+        public virtual void Transmit<T>(in T cmd) where T : struct, IOutgoingCommand
         {
             if (CloseReason != null)
             {
-                lock (_shutdownLock)
+                CheckCanSendWhileClosed(cmd.ProtocolCommandId);
+            }
+
+            Connection.Write(Framing.SerializeToFrames(cmd, ChannelNumber));
```

Comment:
@stebet Here is the conflict that I've mentioned in your PR for the pipelines. 

I get T as input and there is no common type we could use to pass it on to the connection (except the ROM<byte>) without allocation. (If you use the interface IOutgoingCommand then due to T being a struct it will be boxed to to be stored in the interface field, as it doesn't know the actual size of the struct so it stores it as a reference.

Reply:
That's ok, worst case, even though ugly, we could simply create an overload for `Connection.Write` for each possible struct. It's ugly, but it gets the job done and still maintains the benefits of this PR.

Reply:
The issue isn't the Write method itself, it's the channel's storage type (Outputframe vs ROM<byte>). The Outputframe struct can't store the different structs without boxing them => Benefits of this PR are lost.